omniledger/darc/darc.go

```diff
@@ -118,662 +97,745 @@ func NewDarcFromProto(protoDarc []byte) *Darc {
+// using this function.
+func (r Rules) UpdateEvolution(expr expression.Expr) error {
+	return r.updateRule(evolve, expr)
 }
 
-func (r *Request) CopyReq() *Request {
-	rCopy := &Request{
-		DarcID:  r.DarcID,
-		RuleID:  r.RuleID,
-		Message: r.Message,
+func (r Rules) updateRule(a Action, expr expression.Expr) error {
+	if _, ok := r[a]; !ok {
+		return errors.New("action does not exist")
 	}
-	return rCopy
+	r[a] = expr
+	return nil
 }
 
-func (s *Signer) Sign(req *Request) (*Signature, error) {
-	rc := req.CopyReq()
-	b, err := protobuf.Encode(rc)
-	if err != nil {
-		return nil, err
-	}
-	if b == nil {
-		return nil, errors.New("nothing to sign, message is empty")
+func isEvolution(action Action) bool {
+	if action == evolve {
+		return true
 	}
-	if s.Ed25519 != nil {
-		key, err := s.GetPrivate()
-		if err != nil {
-			return nil, errors.New("could not retrieve a private key")
-		}
-		pub, err := s.GetPublic()
-		if err != nil {
-			return nil, errors.New("could not retrieve a public key")
-		}
-		signature, _ := schnorr.Sign(cothority.Suite, key, b)
-		signer := &SubjectPK{Point: pub}
-		return &Signature{Signature: signature, Signer: *signer}, nil
-	}
-	return nil, errors.New("signer is of unknown type")
+	return false
 }
 
-func (s *Signer) SignWithPath(req *Request, path []int) (*SignaturePath, error) {
-	rc := req.CopyReq()
-	b, err := protobuf.Encode(rc)
+// Evolve evolves a darc, the latest valid darc needs to sign the new darc.
+// Only if one of the previous owners signs off on the new darc will it be
+// valid and accepted to sign on behalf of the old darc. The path can be nil
+// unless if the previousOwner is an SignerEd25519 and found directly in the
+// previous darc.
+func (d *Darc) Evolve(path []*Darc, prevOwner *Signer) error {
+	d.Signature = nil
+	prevDarc := path[len(path)-1]
+	d.Version = prevDarc.Version + 1
+	if len(path) == 0 {
+		return errors.New("path should not be empty")
+	}
+	d.BaseID = prevDarc.GetBaseID()
+	sig, err := NewDarcSignature(prevOwner, d.GetID(), path)
 	if err != nil {
-		return nil, err
-	}
-	if b == nil {
-		return nil, errors.New("nothing to sign, message is empty")
-	}
-	if s.Ed25519 != nil {
-		key, err := s.GetPrivate()
-		if err != nil {
-			return nil, errors.New("could not retrieve a private key")
+		return errors.New("error creating a darc signature for evolution: " + err.Error())
+	}
+	if sig == nil {
+		return errors.New("the resulting signature is nil")
+	}
+	d.Signature = sig
+	return nil
+}
+
+// EvolveOnline works like Evolve, but doesn't inlcude all the necessary data
+// to verify the update in an offline setting. This is enough for the use case
+// where the ocs stores all darcs in its internal database.  The service
+// verifying the signature will have to verify if there is a valid path from
+// the previous darc to the signer.
+func (d *Darc) EvolveOnline(prevd *Darc, prevOwner *Signer) error {
+	/*
```

Comment:
Is this dead code and can be removed?

Reply:
it's to remind me that EvolveOnline still needs to be done